pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py

```diff
@@ -0,0 +1,785 @@
+"""
+(*)~---------------------------------------------------------------------------
+Pupil - eye tracking platform
+Copyright (C) 2012-2018 Pupil Labs
+
+Distributed under the terms of the GNU
+Lesser General Public License (LGPL v3.0).
+See COPYING and COPYING.LESSER for license details.
+---------------------------------------------------------------------------~(*)
+"""
+
+import os
+import logging
+import time
+import platform
+import multiprocessing
+import csv
+import itertools
+
+logger = logging.getLogger(__name__)
+if platform.system() == "Darwin":
+    mp = multiprocessing.get_context("fork")
+else:
+    mp = multiprocessing.get_context()
+
+import numpy as np
+import cv2
+import pyglui
+import gl_utils
+import pyglui.cygl.utils as pyglui_utils
+import OpenGL.GL as gl
+
+from plugin import Analysis_Plugin_Base
+import file_methods
+from cache_list import Cache_List
+import player_methods
+
+from surface_tracker.surface_tracker import Surface_Tracker
+from surface_tracker import (
+    offline_utils,
+    background_tasks,
+    Square_Marker_Detection,
+    Heatmap_Mode,
+)
+from surface_tracker.surface_offline import Surface_Offline
+
+
+class Surface_Tracker_Offline(Surface_Tracker, Analysis_Plugin_Base):
+    """
+    - Mostly extends the Surface Tracker with a cache
+    Special version of surface tracker for use with videofile source.
+    It uses a seperate process to search all frames in the world video file for markers.
+     - self.cache is a list containing marker positions for each frame.
+     - self.surfaces[i].cache is a list containing surface positions for each frame
+    Both caches are build up over time. The marker cache is also session persistent.
+    See marker_tracker.py for more info on this marker tracker.
+    """
+
+    def __init__(self, g_pool, marker_min_perimeter=60, inverted_markers=False):
+        self.timeline_line_height = 16
+        self.Surface_Class = Surface_Offline
+        super().__init__(g_pool, marker_min_perimeter, inverted_markers)
```

Comment:
I think the best solution for this would be to make this a property, which is abstract in `Surface_Tracker`. I used capitals in the name because this is a reference to a class rather than an object. If new surfaces are added, they will be initialized using this class. I wanted to reflect this in the name by using the naming convention for classes.